// projects/lib/colors.js
'use strict';

var chalk = require('chalk');
var Color = require('color');
var config = require('./config.js');

var FOREGROUND = 38;
// var BACKGROUND = 48;

var NORMAL = 2;

var CONTROL = '\x1b[';
var RESET = '\x1b[0m';

var DELIMITER = exports.delimiter = ';';

var palette = exports.palette = config.colors.dark['8'];

exports.noop = function (string) {
  return string;
};

exports.colorFn = exports.noop;

exports.color16m = function (string, color) {
  color = new Color(color);

  return CONTROL +
    [
      FOREGROUND,
      NORMAL,
      color.red(),
      color.green(),
      color.blue()
    ].join(DELIMITER) +
    'm' +
    string +
    RESET;
};

// TODO: implement this with Yet Another Color Library
// function color256(string, color) {
//   return string;
// }

exports.color8 = function (string, color) {
  return chalk[color](string);
};

exports.primary = function (string) {
  return exports.colorFn(string, palette.primary);
};

exports.secondary = function (string) {
  return exports.colorFn(string, palette.secondary);
};

exports.highlight = function (string) {
  return exports.colorFn(string, palette.highlight);
};

exports.dim = function (string) {
  return exports.colorFn(string, palette.dim);
};

exports.good = function (string) {
  return exports.colorFn(string, palette.good);
};

exports.neutral = function (string) {
  return exports.colorFn(string, palette.neutral);
};

exports.bad = function (string) {
  return exports.colorFn(string, palette.bad);
};
